### DISCUSSION TITLE
Multiple jobs with same input cause errors due to cache race condition.

### DISCUSSION THREAD

**Discussion by schiotz** [mir-group/nequip#263] [Q&A] [answered]
I am running a number of training runs with the same input.  The job names are specified in the yaml file as
```
root: results/projectname
run_name: YYYY
```
where YYYY is changed in each job.

Unfortunately, pytorch (or NequIP ?) is caching the input file in a folder names ``results/projectname/processed_dataset_HASH``. When a number of jobs are starting simultaneously on the cluster, many of the jobs write to this folder, corrupting the data.  This is some kind of race condition, where the jobs start by seeing that the folder is not there, and then they decide they need to write it.  Others see it being present, and try to read it before it is fully written.

As a workaround, I try to make all but the first job start by sleeping five minutes while the first one makes the cache. Do any of you have any suggestions for how to handle this in a more robust way?

/Jakob

**Comment by Linux-cpp-lisp** [comment, marked as answer]
Hi @schiotz ,

NequIP intentionally caches processed data, which unfortunately can lead to this known race condition that is a little difficult to comprehensively solve without getting into fairly complicated solutions like lock files.

The simple user-level solution is just to pre-pre-process the dataset before running any of the `nequip-train`s. The simplest way to do this is to just run `nequip-benchmark one-of-my-yamls.yaml`, which will trigger preprocessing by loading the data. This command can be run on CPU or login nodes at a cluster as well and does not require a GPU. Then, a cached dataset will exist on disk and you can launch any number of trainings simultaneously without having a problem with the race condition.

**Reply by Linux-cpp-lisp** [reply]
(The "right" solution with file locks is a big pain on networked cluster filesystems, see for example https://stackoverflow.com/questions/48770531/how-to-do-proper-file-locking-on-nfs, and is not something we want to implement just so it can break on every third person's cluster 😁 )

**Reply by schiotz** [reply]
Fair enough!  I'll try to make a hack that works 98% of the time on our particular setup :-)

**Comment by schiotz** [comment]
Thank you, @Linux-cpp-lisp.

I'll do just that, perhaps in a job the other jobs depend on.
